src/coreclr/jit/gtlist.h

```diff
@@ -242,6 +242,11 @@ GTNODE(SETCC            , GenTreeCC          ,0,GTK_LEAF|DBK_NOTHIR)
 // The XARCH BT instruction. Like CMP, this sets the condition flags (CF to be precise) and does not produce a value.
 GTNODE(BT               , GenTreeOp          ,0,(GTK_BINOP|GTK_NOVALUE|DBK_NOTHIR))
 #endif
+// Sets the condition flags according to the combined results of its children.
+#if defined(TARGET_ARM64)
+GTNODE(CCMP_EQ          , GenTreeOp          ,0,GTK_BINOP|DBK_NOTHIR)
+GTNODE(CCMP_NE          , GenTreeOp          ,0,GTK_BINOP|DBK_NOTHIR)
```

Comment:
Once we switch compare chains to this scheme I assume we will need the rest of the conditions too, right? So we will probably end up with some `GenTreeCCMP` that is a binop with an additional `GenCondition` in it.

Reply:
The way I've done it is that `CCMP_EQ(LT,NE)` is essentially identical to `NE(AND(LT,NE),0)`. Which is generated as:
```
cmp
ccmp LT
b NE
```
The CCMP_EQ itself doesn't generate any code, just ensures the `CMP`s become `CCMP`s.

Are you suggesting that instead we have the `CCMP` nodes behave in a similar way to `CMP` nodes? The above code would be `CCMP_LT(NE)`. The `CCMP_LT` would be a three Op node similar to SELECT (previous condition + 2 operands).
Assuming there are no compare specific optimizations happening post lowering (which I don't think there is) then that would be ok.

Reply:
> Are you suggesting that instead we have the CCMP nodes behave in a similar way to CMP nodes?

Yes, with `GTF_SET_FLAGS` being set. That is, CCMP would correspond directly to the `ccmp` being generated. So we would end up lowering something like:
```scala
l1 = ...
l2 = ...
l = LT(lt1, lt2)
n1 = ...
n2 = ...
n = NE(n1, n2)
a = AND(l, n)
e1 = CNS_INT 0
e = NE(a, e1)
JTRUE(e, ...)
```
into (with necessary reordering checks):
```scala
l1 = ...
l2 = ...
n1 = ...
n2 = ...
LT(l1, l2) [with GTF_SET_FLAGS; now doesn't produce a value, and is actually just a "CMP"]
CCMP(n1, n2, LT, zero-flag) [with GTF_SET_FLAGS]
JCC(NE)
```
The existing lowering could first change `NE(AND(x, y), 0)` into `TEST_NE(x, y)`, if that makes it easier to recognize.

SELECT would get a variant that has a CC and consumes from flags (this already kind of exists on x86, as `SELECT_HI`). I think this is more similar to the initial implementation you had, except for the reordering that we need to take care of.

If the consumer cannot use the value directly from flags then we would insert a `SETCC` like we do in some other cases today.

In the future, once we make flag dependencies more explicitly, we would have some edges between these nodes to indicate those dependencies.

Anyway, if you don't want to do this work as part of this I completely understand -- I think the approach you went with here is ok and we can switch it to this at some later point in time. Something like this will be needed to implement the 'harder' cases in the xarch backend, so at some point I will probably do this work.

Reply:
FWIW, I'm doing some preparatory changes in #82020 to be able to make these changes.

Reply:
I've been working through adding the three operand GT_CCMP_EQ/NE/LT/LE... nodes. Still working through some asserts failures (and various distractions along the way).
It feels a lot neater too. It's no longer dependant on lots of AND chain parsing - all of that code just stays with the boolean ANDs

A quick glance at https://github.com/dotnet/runtime/pull/82020 and I think it's ok with my code, but I'll take a closer look.